src/coloring.c

```diff
@@ -0,0 +1,107 @@
+
+#include "igraph_coloring.h"
+#include "igraph_interface.h"
+#include "igraph_adjlist.h"
+#include "igraph_interrupt_internal.h"
+#include <assert.h>
+
+
+int igraph_i_vertex_coloring_greedy_cn(const igraph_t *graph, igraph_vector_int_t *colors) {
+    igraph_integer_t i, vertex, maxdeg;
+    igraph_integer_t vc = igraph_vcount(graph);
+    igraph_vector_int_t cn; /* number of already coloured neighbours */
+    igraph_adjlist_t adjlist;
+
+    IGRAPH_CHECK(igraph_vector_int_resize(colors, vc));
+    igraph_vector_int_fill(colors, 0);
+
+    IGRAPH_CHECK(igraph_vector_int_init(&cn, vc));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &cn);
+
+    IGRAPH_CHECK(igraph_adjlist_init(graph, &adjlist, IGRAPH_ALL));
+    IGRAPH_FINALLY(igraph_adjlist_destroy, &adjlist);
+
+    /* find maximum degree and a corresponding vertex */
+    {
+        igraph_vector_t degree;
+
+        IGRAPH_CHECK(igraph_vector_init(&degree, 0));
+        IGRAPH_FINALLY(igraph_vector_destroy, &degree);
+        IGRAPH_CHECK(igraph_degree(graph, &degree, igraph_vss_all(), IGRAPH_ALL, 0));
+
+        vertex = (igraph_integer_t) igraph_vector_which_max(&degree);
+        maxdeg = VECTOR(degree)[vertex];
+
+        igraph_vector_destroy(&degree);
+        IGRAPH_FINALLY_CLEAN(1);
+    }
+
+    while (VECTOR(cn)[vertex] >= 0) {
+        igraph_vector_int_t *neighbors = igraph_adjlist_get(&adjlist, vertex);
+        long neigh_count = igraph_vector_int_size(neighbors);
+
+        /* colour current vertex */
+        {
+            igraph_integer_t col;
+            igraph_vector_int_t neigh_colors;
+
+            IGRAPH_CHECK(igraph_vector_int_init(&neigh_colors, neigh_count));
```

Comment:
It is probably unfortunate to re-allocate `neigh_colors` in every iteration of the loop. I would move the allocation outside the `while` loop and just use `igraph_vector_int_resize()` here.

Note that if the allocation is moved outside the `while` loop, you'll also need to guard it with `IGRAPH_FINALLY(igraph_vector_int_destroy, &neigh_colors)`.

Reply:
Fixed, though it didn't result in an appreciable speedup.